File: 2021-blind/Q3_1.py
import bisect

l1 = ['-', 'cpp', 'java', 'python']
l2 = ['-', 'backend', 'frontend']
l3 = ['-', 'junior', 'senior']
l4 = ['-', 'chicken', 'pizza']

def solution(info, query):
    # A. 전처리
    L = [[] for i in range(108)]
    for s in info:
        z = s.split(' ')
        v1, v2, v3, v4 = l1.index(z[0]), l2.index(z[1]), l3.index(z[2]), l4.index(z[3])
        v5 = int(z[4])
        for c1 in [0, v1]:
            for c2 in [0, v2]:
                for c3 in [0, v3]:
                    for c4 in [0, v4]:
                        idx = c1*27 + c2*9 + c3*3 + c4
                        L[idx].append(v5)

    for i in range(108):
        L[i].sort()
    #B. 쿼리
    ret = []
    for q in query:
        z = q.split(' ')
        v1, v2, v3, v4 = l1.index(z[0]), l2.index(z[2]), l3.index(z[4]), l4.index(z[6])
        target = int(z[7])
        idx = v1*27 + v2*9 + v3*3 + v4
        ret.append(len(L[idx]) - bisect.bisect_left(L[idx], target))
    return ret

File: 2021-blind/Q3_1.java
import java.util.*;

class Solution {
    public int[] solution(String[] info, String[] query) {
        String[][] l = {{"-", "cpp", "java", "python"},
                          {"-", "backend", "frontend"},
                          {"-", "junior", "senior"},
                          {"-", "chicken", "pizza"}};
        // A. 전처리
        ArrayList<Integer>[] L = new ArrayList[108];
        for(int i = 0; i < 108; i++)
            L[i] = new ArrayList<Integer>();
        for(int t = 0; t < info.length; t++){
            String s = info[t];
            int v[] = new int[4];
            int idx1 = 0, idx2 = s.indexOf(" ");
            for(int i = 0; i < 4; i++){
                String cond = s.substring(idx1, idx2);
                idx1 = idx2+1;
                idx2 = s.indexOf(" ", idx1);
                List<String> tmp = Arrays.asList(l[i]);
                v[i] = tmp.indexOf(cond);
            }
            int score = Integer.parseInt(s.substring(idx1, s.length()));
            int factor[] = {27, 9, 3, 1};
            for(int i = 0; i < 16; i++){
                int idx = 0;
                for(int j = 0; j < 4; j++){
                    if((i & (1<<j)) != 0) idx += factor[j] * v[j];
                }
                L[idx].add(score);
            }
        }
        for(int i = 0; i < 108; i++) Collections.sort(L[i]);
        // B. 쿼리
        int[] ans = new int[query.length];
        for(int t = 0; t < query.length; t++){
            String q = query[t];
            int v[] = new int[4];
            int idx1 = 0, idx2 = q.indexOf(" ");
            for(int i = 0; i < 4; i++){
                String cond = q.substring(idx1, idx2);
                idx1 = idx2+5;
                idx2 = q.indexOf(" ", idx1);
                List<String> tmp = Arrays.asList(l[i]);
                v[i] = tmp.indexOf(cond);
            }
            int target = Integer.parseInt(q.substring(idx1 - 4, q.length()));
            int idx = v[0]*27 + v[1]*9 + v[2]*3 + v[3];
            int st = 0, en = L[idx].size();
            while(st < en){
                int mid = (st+en) / 2;
                if(L[idx].get(mid) >= target) en = mid;
                else st = mid+1;
            }
            ans[t] = L[idx].size() - st;
        }
        return ans;
    }
}
